Feature/user interface

MrSayyod opened the pull request:
Due to requirements:
- followed the GitFlow guidelines;
- created content for the user interface;
- created basic flow.
In this step main.rb file only displays users' names, their choices and game board.

MrSayyod committed:
added initial steps

Hermela-Fikre committed:
Added the board class and display_board method

MrSayyod committed:
added winning and checking methods

Hermela-Fikre committed:
Fix rubocop errors

mosesogwo reviewed (changes requested):
**HI @MrSayyod, @Hermela-Fikre:clap:. You are off to a good start in this milestone:+1:. However, there are some requirements that have not been satisfied.**
- **Your current code does not show the flow of the game at all. You should show the following behaviors in your game:**
1. game tells which player turn it is
2. game asks player to select from available moves
3. game informs player if selected move is invalid
4. game displays board after player move
5. game informs player if selected move is a winning move
6. game informs player if selected move is a draw move
7. game repeats all actions for next player's move

The above actions should be repeated for each player turn, so the correct implementation should have a loop that repeats these steps for each player until there is a win or a draw.

It doesn't have to work completely, but you have to show that you have thought about all these in your code.

![](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone2/logic.png)


_**Kindly make these changes and submit another code review request.**_

**Any questions? You can reach me on Slack (Moses Ogwo)**

**Cheers:stars:**

### Status: Changes Requested.

MrSayyod committed:
added turn and checking methods

MrSayyod commented:
Added turn and checking methods. 
Turn method tells players' turns and asks to select available moves. If player chose a non-available character it it says invalid character and repeats asking to select valid characters. after each turn board displays.
The checking method checks if the chosen cell is taken or not.

Hermela-Fikre committed:
Added the winning,draw  method

MrSayyod committed:
fixed linter errors

MrSayyod commented:
All requirements have been done)

stratospherique reviewed (changes requested):
### Great work implementing some of the suggested changes :clap: However, some changes need to be applied in order to get approval for this milestone. Check :arrow_down: 

- **[REQUIRED]** **main.rb** needs to be executable. :pray: fix that.

- **[REQUIRED]** The game execution stops after gathering user input. :pray: fix it
![image](https://user-images.githubusercontent.com/26124023/80605785-6d49c780-8a2b-11ea-89b5-422e8bb56232.png)

- **[OPTIONAL BUT REQUIRED FOR NEXT MILESTONE]** Some of the requirements for the next milestone are that all classes are kept in /lib folder and that /lib folder shouldn't contain any UI related code (e.g puts, gets, etc...). Thus, you should refactor your Board class so it doesn't contain UI code. you could make UI related code as global methods(e.g display board, display game status and result, etc...) in **main.rb** file.

- **[REQUIRED]** The game loop is not DRY enough, you should implement a turn switch mechanism so that you always enter the loop with the current player and you don't have to repeat instructions for both players in a single iteration.

- **[SUGGESTION]** Since your game loop (i.e turn method) contains UI related code, I suggest that you move it outside of the Board class because it will be hard to separate UI related code from it if you decided to keep the turn method.

### That's it :relieved: When finished, kindly submit a code re-review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

MrSayyod committed:
added combination method

MrSayyod committed:
added current_player method

Hermela-Fikre committed:
Fix rubocob errors and created board file on lib folder

Hermela-Fikre committed:
Created board file in the lib folder

MrSayyod committed:
fixed linter errors

MrSayyod commented:
- main.rb is executable;
- game execution stopping fixed;
- class Board refactored;
- turn method which contains UI codes is out of class Board now;
- DRY codes reduced.

shshamim63 commented on bin/main.rb:
```diff
@@ -1,2 +1,42 @@
 #!/usr/bin/env ruby
+require '../lib/board.rb'
 puts 'Hello world!'
```
Is it needed?

MrSayyod replied:
if you mean "Hello world!", it is from milestone 1 and we didn't delete it.

MrSayyod replied:
deleted it)

shshamim63 commented on bin/main.rb:
```diff
@@ -1,2 +1,42 @@
 #!/usr/bin/env ruby
+require '../lib/board.rb'
```
whenever you require a file always use `require_relative`.

MrSayyod replied:
Ok 👍  Thanks. But can you tell us what is the difference between require and require_relative? To prevent this error for the next time it is good to know the difference)))

whiz25 replied:
@MrSayyod `require_relative` is used to load modules locally while `require` is used to load global modules. IE modules that are installed as gems or modules that are available by default in ruby

shshamim63 reviewed (changes requested):
## Status :1234: Required Changes :1234: 
___
Hi @MrSayyod and @Hermela-Fikre :wave: ,
Awesome job :clap: . You are almost there. Let me just share what is actually the aim of this milestone. That is, ensuring that you know what will be the process/steps of this game. They are no need that everything should be workable(just it should be executable). So what are the main steps?
**Thing you have in your main.rb**
- asks players for name[optional]
- display the board
- informs which players turn is it.
- asks for a move
- informs if the move is invalid.
- contains a loop to indicate that you know there are repetitive tasks.
**Case you don't have in your main.rb**
- Inform player if he wins
- Infor players if it is a draw.
[You can cover it for this milestone with comment].
####  Now the task you have to cover to get an approval for this milestone
1. Remove the hello world print.
2. Replace require with require_relative.
3. Adds two messages related to declaring a winner or if it is a draw.
___
##### Please fix the above-mentioned issue and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain

MrSayyod committed:
added messages about winning, deleted 'hello world' and 'require' replaced with 'require_relative'

MrSayyod commented:
- removed the `hello world ` print.
- replaced `require` with `require_relative`.
- added two messages related to declaring a winner or if it is a draw.

whiz25 reviewed (changes requested):
## Nice Progress @MrSayyod 
You have made some of the requested changes by the previous reviewers. We need to fix a few things regarding this PR before we can merge.

- [x] add logic for informing player if he wins.
- [x] add logic to inform players if it is a draw.
- [x] make the suggested changes and resubmit for a review

### Martin Afani

whiz25 commented on bin/main.rb:
```diff
@@ -1,2 +1,41 @@
+puts "Welcome to the game Tic Tac Toe, #{player1_name} and #{player2_name}! #{player1_name} plays with #{player1_select} and #{player2_name} plays with #{player2_select}. You can start now. Happy gaming)" # rubocop:disable Layout/LineLength
+
+def turn
+  until array.none?('') == false
```
`array.none?(' ')` returns `true` by default as a boolean. So you can do this:
`if array.none?(' ')
    # handle true outcome
else
    # handle false outcome
end`

MrSayyod replied:
if I use array.none?('') it is going false, not true as you said
p.s. ('') without space between quotes

MrSayyod replied:
if it will be with space, then it comes `true`

whiz25 replied:
@MrSayyod I meant `array.none?('')` without the space.

MrSayyod replied:
@whiz25

MrSayyod replied:
and my all array items are ''(without space between quotes) not ' '(space between quotes). Therefore i checked my array with ''''(without space between quotes)

MrSayyod replied:
when i use if statement it will not loop, won't it?

whiz25 replied:
If the `if statement is inside the loop, then it will loop`

MrSayyod replied:
`until array.none?('') == false`
    `if array.none?('')`
like this

whiz25 commented on bin/main.rb:
```diff
@@ -1,2 +1,41 @@
+  until array.none?('') == false
+    puts "It is your turn #{current_player[0]}. Please, choose numbers between 1 and 9."
+    current_player[2] = gets.chomp.to_i
+    if (1..9).include?(current_player[2]) == true
```
You can do this: 
`if (1..9).include?(current_player[2])`
    # handle true outcome
`else`

MrSayyod replied:
Do you mean no need for `true` boolean?

whiz25 replied:
@MrSayyod yes

MrSayyod replied:
@whiz25

MrSayyod replied:
👍

MrSayyod committed:
changed some codes due to TSE review

MrSayyod committed:
fixed linter errors

MrSayyod commented:
The logic for informing the player about win or draw is in the `combination` method in the lib/ folder.
The `turn` method changed:
- added `next unless array.none?('')`
- removed `true` boolean from `if (1..9).include?(current_player[2]) == true`

Anna-Myzukina commented:
## Hi team ! Very nice start! :wave: 

# [APPROVED]  :+1:


## [OPTIONAL]
` You can think abut next changes: `

* [x] Using `require_relative` for files in your directory instead of `require` is a ruby convention.

* [x] Use attr_readers/attr_writers instead of attr_accessors. Be aware that by using it they allow writing that property from outside the class. That variable can be set to anything (from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

![screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/attr_accessor.png) 

### Go and merge to develop. :100: 
`Goodluck in your next milestones Happy coding ` <g-emoji class="g-emoji" alias="computer" fallback-src="https://github.githubassets.com/images/icons/emoji/unicode/1f4bb.png"><img class="emoji" alt="computer" height="25" width="25" src="https://github.githubassets.com/images/icons/emoji/unicode/1f4bb.png"></g-emoji>
